theano/tensor/opt.py

```diff
@@ -6085,6 +6085,45 @@ def local_log_add(node):
                 return [ret]
 
 
+@gof.local_optimizer([T.log])
+def local_log_sum_exp(node):
+    # log(sum_i(exp(x_i))) = x_max + log(sum_i(exp(x_i - x_max)))
+
+    if node.op != T.log:
+        return
+
+    sum_node = node.inputs[0].owner
+    # If the sum has keepdims=True, there might be a dimshuffle
+    if sum_node and isinstance(sum_node.op, T.DimShuffle):
+        sum_node = sum_node.inputs[0].owner
+
+    if not sum_node or not isinstance(sum_node.op, T.Sum):
+        return
+
+    exp_node, axis = sum_node.inputs[0].owner, sum_node.op.axis
+    if not exp_node or not (
+            isinstance(exp_node.op, Elemwise) and
+            isinstance(exp_node.op.scalar_op, scalar.Exp)):
+        return
+
+    pre_exp = exp_node.inputs[0]
+    max_pre_keepdims = T.max(pre_exp, axis=axis, keepdims=True)
+
+    ret = (max_pre_keepdims + T.log(T.sum(T.exp(pre_exp - max_pre_keepdims),
+                                          axis=axis, keepdims=True)))
+
+    # Restore shape and broadcastable pattern
+    ret = T.reshape(ret, node.inputs[0].shape)
```

Comment:
This line is problematic:
- if the initial `sum_node.op` was Dimshuffle and it had actually transposed the sum, then a Reshape does not do the right thing, we should apply a Dimshuffle again instead
- depending on `node.inputs[0]` is a bad idea, because it re-introduces part of the old graph in the new one.